benches/src/bin/collect.rs

```diff
@@ -596,7 +596,127 @@ fn slope(x_y: Vec<(u64, u64)>) -> f64 {
         .map(|(x, y)| (*x as f64 - avg_x) * (*y as f64 - avg_y))
         .sum();
     let sq_x: f64 = x_y.iter().map(|(x, _)| (*x as f64 - avg_x).powi(2)).sum();
-    sum_x_y / sq_x
+
+    // The original formula says:
+    //
+    // y = B * x,
+    // where B = sum((x_i - x_avg) * (y_i - y_avg)) / sum((x_i - x_avg) ^ 2) = sum_x_y / sq_x
+    //
+    // We want to know how many elements fit into one `noop` opcode timing,
+    // so we need value of `x / y`:
+    //
+    // y = B * x
+    //     |
+    //    \|/
+    // 1 / B = x / y
+    //     |
+    //    \|/
+    // 1 / sum_x_y / sq_x = x / y
+    //     |
+    //    \|/
+    // sq_x / sum_x_y = x / y
+    sq_x / sum_x_y
+}
+
+fn dependent_cost(x_y: Vec<(u64, u64)>) -> (u64, u64) {
+    const NEAR_LINEAR: f64 = 0.1;
+
+    enum Type {
+        /// The points have a linear property. The first point
+        /// and the last points are almost the same(The difference is < 0.1).
+        Linear,
+        /// When the delta of the last point is much lower than
+        /// the first point, it is a logarithmic chart.
+        Logarithm,
+        /// When the delta of the last point is much more than
+        /// the first point, it is an exponential chart.
+        Exp,
+    }
+
+    #[derive(Clone, Copy)]
+    struct Point {
+        /// Number of elements for the opcode.
+        x: u64,
+        /// Time in `noop`s required to process `x` elements.
+        ///
+        /// Note: If the time to process `noop` opcode is `20` nanoseconds
+        /// and the time required to process `x` elements is `140`,
+        /// the `y` is 7 in this case.
+        y: u64,
+    }
+
+    impl Point {
+        /// The price in `noop` time to process a single element.
+        fn price(&self) -> f64 {
+            (self.y as f64) / (self.x as f64)
+        }
+
+        /// The amount that this operation costs per increase in one `noop` time.
+        fn amount(&self) -> f64 {
+            (self.x as f64) / (self.y as f64)
+        }
+    }
+
+    let linear_regression = linear_regression(x_y.clone());
+
+    let x_y = x_y
+        .into_iter()
+        .map(|(x, y)| Point { x, y })
+        .collect::<Vec<_>>();
+
+    let first = x_y.first().unwrap();
+    let last = x_y.last().unwrap();
+
+    let near_linear = linear_regression * NEAR_LINEAR;
+    let expected_type = if (linear_regression - first.amount()).abs() < near_linear
```

Comment:
A more sound approach for testing the goodness-of-fit using linear regression would be the $R^2$ metric, or [coefficient of determination](https://en.wikipedia.org/wiki/Coefficient_of_determination#Definitions). Where 1 represents a perfect fit, and lower values indicate a poor fit of the linear model.

![image](https://github.com/FuelLabs/fuel-core/assets/794823/1c2e4151-ca68-447e-9bbb-349667f4bb52)

Reply:
We could use it here, but we don't have cases of linear behavior, so the super-complex formula is overkill. Because each opcode includes the base execution time, it looks like a logarithmic almost in all cases.

Plus, we still need to play with the threshold for `R`. The code for the logarithm behavior should cover almost all cases, and the `min` strategy means we can only overcharge user. If later, with benchmarking target block time, we will see that some opcodes are really overcharged, then we can try to improve the algorithm. But right now, it looks like our benchmarks produce undercharge costs.